Add full clone setting for prebuilds

mustard-mh opened the pull request:
## Description
<!-- Describe your changes in detail -->
Improve large repository workspace startup performance by fully clone repository in prebuilds phase (download + unzip speed is faster than SCM processing)

## Related Issue(s)
<!-- List the issue(s) this PR solves -->
Fixes ENT-730

## How to test
<!-- Provide steps to test this PR -->
- Add `SUPERVISOR_DEBUG_ENABLE=true` to preview env's environments https://hw-large-repo.preview.gitpod-dev.com/user/variables so that we get debug logs from supervisor
- Create workspace with context https://github.com/JetBrains/intellij-community
   - 1-a) Check log of `ws-daemon` > `clone repository took`
   - 1-c) Check logs of workspace > `unshallow of local repository took`
- Create project for https://github.com/JetBrains/intellij-community, and enable prebuild (only enable), trigger prebuild and create a workspace after prebuild succeed
   - 2-a) Check logs of `ws-daemon`  > `clone repository took`
   - 2-b) Check logs of `ws-daemon` > `aria2c download duration`, `extract tarbal duration`
   - 2-c) Check logs of workspace > `unshallow of local repository took`
- Enable full clone of project in project settings page, trigger prebuild in UI and create a workspace after prebuild succeed
   - 3-a) Check logs of `ws-daemon`  > `clone repository took`
   - 3-b) Check logs of `ws-daemon` > `aria2c download duration`, `extract tarbal duration`
   - 3-c) Check logs of workspace > `unshallow of local repository took`
- Duration of `3-b + 3-c` should less than `1-a + 1-c` and `2-a + 2-b + 2-c`
- Workspace 3 git functions should works like workspace 1 /2 (i.e. `git lg`)

✅ Shared test result

| Duration | |Workspace 1|WS 2 | WS 3|
|:-:|:-:|:-:|:-:|:-:|
|||pure workspace|prebuild + workspace|prebuild(full clone) + workspace|
|**a)** |ws-daemon git clone|52.920714931s| `prebuild` 49.110767306s| `prebuild` 4m50.046374877s|
|**b)** | ws-daemon download and extract tar|NaN|8.946159792s+18.608708639s|26.825172768s+32.283299312s|
|||`accessable`|`accessable`|`accessable`|
|**c)** | supervisor fetch|7m35.843093833s|7m0.284969917s|2.52514ms|
|summary|| `access=53s`, `performance=7m35s`|`access=27.6s`, `performance=7m`|`access=59s`, `performance=0s`|

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - hw-large-repo</li>
	<li><b>🔗 URL</b> - <a href="https://hw-large-repo.preview.gitpod-dev.com/workspaces" target="_blank">hw-large-repo.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - hw-large-repo-gha.28542</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-hw-large-repo%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-dev-preview" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [x] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [x] /werft preemptible
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

|Dashboard UI Changes|
|:-:|
|<img width="604" alt="image" src="https://github.com/user-attachments/assets/638fe67c-6d31-4d07-a156-5965a518a4a1">|

/hold

mustard-mh commented:
✅ Shared test result

| Duration | |Workspace 1|WS 2 | WS 3|
|:-:|:-:|:-:|:-:|:-:|
|||pure workspace|prebuild + workspace|prebuild(full clone) + workspace|
|**a)** |ws-daemon git clone|52.920714931s| `prebuild` 49.110767306s| `prebuild` 4m50.046374877s|
|**b)** | ws-daemon download and extract tar|NaN|8.946159792s+18.608708639s|26.825172768s+32.283299312s|
|||`accessable`|`accessable`|`accessable`|
|**c)** | supervisor fetch|7m35.843093833s|7m0.284969917s|46.557682254s|
|summary|| `access=53s`, `performance=7m35s`|`access=27.6s`, `performance=7m`|`access=59s`, `performance=46s`|

filiptronicek commented on components/public-api/gitpod/v1/configuration.proto:
```diff
@@ -33,6 +33,8 @@ message PrebuildSettings {
   string workspace_class = 5;
   PrebuildTriggerStrategy trigger_strategy = 6;
+  // full_clone determines if the entire repository should be cloned.
+  bool full_clone = 7;
```
I kind of wish we had something like `ClonePreferences` wrapping this, so that we don't just pollute `PrebuildSettings`. Maybe it's not worth it, and we'll never even think about adding another prop in there, but topologically it makes sense to me, and it adds some kind of future proofing without the potential need to break this prop.

filiptronicek replied:
Not necessarily; to me, both `workspace_class` is not worth abstracting any further and `branch_matching_pattern` I must say I kind of wish it was under a `PrebuildTrigger` message, which would also include `branch_strategy` and `trigger_strategy`.

mustard-mh replied:
I don't understand, do you mean to use `CloneSettings` instead of `bool`? 
```
message CloneSettings {
  UNSPECIFIED
  DEPTH_ONE
  FULL_CLONE
} 
```

If so, we tested before start work on this PR that, both `depth (ws-daemon) + fetch (supervisor)` and `depth (ws-daemon) + fetch depth (supervisor)` are slower than `fully clone (ws-daemon)`

filiptronicek replied:
Sorry for the confusion here, I mean that we'd essentially wrap `full_clone` in a `CloneSettings` message, to which we could potentially add more setings down the line.

filiptronicek replied:
like 
```protobuf
message CloneSettings {
  full_clone bool = 1;
}
```

mustard-mh replied:
yeah, that's what I mean, it's the same case with `workspace_class` that there is no other options we could set.

And if we want to re-struct this message, may make combine `clone_setting` (full_clone) and `workspace_class` together as `policies` looks more align to what we have in organization settings? because they should be in the same layer -> how we configure a prebuild workspace

```
message PrebuildPolicies {
  string workspace_class = 5;
  bool full_clone = 7;
}
message PrebuildTrigger {
  PrebuildTriggerStrategy trigger_strategy = 6;
  string branch_matching_pattern = 2;
  BranchMatchingStrategy branch_strategy = 3;
  int32 prebuild_interval = 4;
}
```

mustard-mh replied:
Personally, I don't like to have lots of message layers, but use enum instead of bool would be more clear

geropl replied:
Uh, good idea!

geropl commented on components/content-service-api/initializer.proto:
```diff
@@ -61,6 +61,9 @@ message GitInitializer {
     // config specifies the Git configuration for this workspace
     GitConfig config = 6;
+
+    // full_clone determines if the entire repository should be cloned
```
```suggestion
    // full_clone determines if the entire repository should be cloned, instead of with `--depth=1`
```

geropl commented on components/content-service-api/typescript/src/initializer_grpc_pb.js:
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
```
@mustard-mh Could you try to remove the copyright date changes from the PR? Ideally those are never touched after the file got created, plus it spams the diff. :pray:

filiptronicek commented on components/content-service/pkg/git/git.go:
```diff
@@ -338,8 +342,18 @@ func (c *Client) Clone(ctx context.Context) (err error) {
+		log.WithField("duration", time.Since(now).String()).WithField("FullClone", c.FullClone).Info("clone repository took")
+	}()
+
 	args := []string{"--depth=1", "--shallow-submodules", c.RemoteURI}
 
+	if c.FullClone {
+		args = []string{c.RemoteURI}
+	}
```
nit: here I think it'd be more readable to do the inverse (i.e. start with `c.RemoteURI` and check `!c.FullClone`).

mustard-mh replied:
I would prefer to keep old logic first

filiptronicek replied:
Am fine with both, resolving

geropl commented on components/content-service/pkg/initializer/git.go:
```diff
@@ -235,8 +250,13 @@ func (ws *GitInitializer) realizeCloneTarget(ctx context.Context) (err error) {
 		// and deal with any error appropriately (i.e. emit a warning rather than fail).
-		if err := ws.Git(ctx, "fetch", "--depth=1", "origin", "--recurse-submodules=no", ws.CloneTarget); err != nil {
-			log.WithError(err).WithField("remoteURI", ws.RemoteURI).WithField("branch", ws.CloneTarget).Error("Cannot fetch remote branch")
+		fetchArgs := []string{"--depth=1", "origin", "--recurse-submodules=no", ws.CloneTarget}
```
:+1:

mustard-mh commented on components/content-service/pkg/initializer/git.go:
```diff
@@ -199,6 +200,20 @@ func (ws *GitInitializer) Run(ctx context.Context, mappings []archive.IDMapping)
 }
 
+func (ws *GitInitializer) isShallowRepository(ctx context.Context) bool {
+	out, err := ws.GitWithOutput(ctx, nil, "rev-parse", "--is-shallow-repository")
```
Copied from supervisor

mustard-mh commented:
I will address feedback later

geropl commented:
Thank you for the nice PR @mustard-mh ! :pray:
I think this is a good solution to the "optimize clone time with prebuilds" usecase. We should definitely roll this out, and if this is helpful for Dynatrace.

One aspect is not covered, though, which is the no-prebuild case: There, we still require the full ~7mins for the "unshallow", where SCM extensions are not usable, because the `.git` folder is locked - which was the original issue the ticket was created for.

How do you think we could tackle that issue, especially in the context of this change:
 - would it make sense to add another flag/option to enable an "incremental unshallow" strategy in supervisor right away?
   - or should we explore that in a follow-up?
 - should we create a separate issue for that?

geropl commented:
Code mostly LGTM, I like [this suggestion](https://github.com/gitpod-io/gitpod/pull/20177#discussion_r1751372092), the rest if more nits/minor suggestions. :heavy_check_mark: 

I tried to test, but somehow my `SUPERVISOR_DEBUG_ENABLE` env var is not picked up, so I don't get proper log output from the workspace (but also, am trusting you that it's fine). :smile:

mustard-mh commented:
> I tried to test, but somehow my SUPERVISOR_DEBUG_ENABLE env var is not picked up, so I don't get proper log output from the workspace (but also, am trusting you that it's fine). 😄

@geropl  It takes time to clone, you may need to open workspace and wait for 10 minutes to get the log

mustard-mh commented:
> Thank you for the nice PR @mustard-mh ! 🙏 I think this is a good solution to the "optimize clone time with prebuilds" usecase. We should definitely roll this out, and if this is helpful for Dynatrace.
> 
> One aspect is not covered, though, which is the no-prebuild case: There, we still require the full ~7mins for the "unshallow", where SCM extensions are not usable, because the `.git` folder is locked - which was the original issue the ticket was created for.
> 
> How do you think we could tackle that issue, especially in the context of this change:
> 
> * would it make sense to add another flag/option to enable an "incremental unshallow" strategy in supervisor right away?
>   
>   * or should we explore that in a follow-up?
> * should we create a separate issue for that?

@geropl  I thought we don't need to deal with no-prebuild case

1. `prebuids` are used to pre-configure your workspaces to save your time, so `full clone` is similar to that tasks init (e.g. install npm packages) does: prepare your workspace and make it ready to code
2. This PR provides a proper workaround for large repository
3. We [tested](https://linear.app/gitpod/issue/ENT-730/unshallow-clone-causes-performance-issues-with-large-repositories#comment-7aeddae8) to fetch with depth does not save time
4. An incomplete git history can break git functionalities: i.e. I used to do `git rebase -i HEAD~10` once workspace startup, and it rebased everything because we clone with depth=1

geropl commented:
> @geropl I thought we don't need to deal with no-prebuild case

We have, at least in the wider context, because we need to support opening big(ger) repositories, without blocking SCM integrations. Let's tackle that in a separate PR. :+1:

mustard-mh commented:
Double checked result is similar after addressing feedbacks

<img width="876" alt="image" src="https://github.com/user-attachments/assets/759b4c29-80a3-4ebf-af4d-d9eeff130949">

geropl reviewed (approved):
Sorry for the delay:

Code LGTM, tested and works! :heavy_check_mark: :muscle:
